Send notifications from the UI process to the WebContent process

pvollan opened the pull request:
#### f4bb61eb4b5e218f8b4f76ac94f753b764a13bac
<pre>
Send notifications from the UI process to the WebContent process
<a href="https://bugs.webkit.org/show_bug.cgi?id=263830">https://bugs.webkit.org/show_bug.cgi?id=263830</a>
<a href="https://rdar.apple.com/117629355">rdar://117629355</a>

Reviewed by Brent Fulgham.

Send notifications from the UI process to the WebContent process. This is to prepare for blocking of notifyd in the WebContent process sandbox.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::postNotification):
(WebKit::WebProcess::postObserverNotification):

Canonical link: <a href="https://commits.webkit.org/270030@main">https://commits.webkit.org/270030@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c79fd79ca28d0505a3258024f972af3598c4b980

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25351 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26401 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22341 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4015 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24735 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22785 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24513 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1897 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20970 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26990 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1643 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28115 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22139 "Found 1 new API test failure: TestWebKitAPI.ServiceWorker.WindowClientNavigate (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22194 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25900 "Found 2 new API test failures: /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/dom-cache, /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/origin-and-total-storage-ratio (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1580 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/19239 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/2858 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5827 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1980 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1946 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -645,6 +645,42 @@ static void logProcessPoolState(const WebProcessPool& pool)
     m_weakObserver = adoptNS([[WKProcessPoolWeakObserver alloc] initWithWeakPtr:*this]);
 
+#if ENABLE(NOTIFYD_BLOCKING_IN_WEBCONTENT)
+    const char* messages[] = { "com.apple.WebKit.LibraryPathDiagnostics",
```
I wonder if this should live in its own `#include` file, so that we could easily add new messages to it?

And are these identical to what's needed on iOS, or might different messages be needed?

cdumez replied:
Would be nice to use `ASCIILiteral`

brentfulgham replied:
Could this be `constexpr Vector<ASCIILiteral> messages { "com.apple.WebKit.LibraryPathDiagnostics", ...etc...`?

brentfulgham replied:
I can't remember if that initialization works.

cdumez replied:
If we're going to use a static vector, then we may as well store Strings inside the Vector since that'd what we end up needing. Otherwise, we keep converting from const char* (or ASCIILiteral) to String to send the IPC.

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -645,6 +645,42 @@ static void logProcessPoolState(const WebProcessPool& pool)
+                                "com.apple.WebKit.showAllDocuments",
+                                "com.apple.WebKit.logMemStats",
+                                "com.apple.WebKit.logPageState" };
+    m_notifyTokens.reserveInitialCapacity(std::size(messages));
```
I think this could be rewritten as something like:

```
m_notifyTokens = WTF::map(messages, [](const char* message) -> int {
    int notifyToken;
    notify_register_dispatch(message, &notifyToken, dispatch_get_main_queue(), ^(int token) {
        for (auto& process : m_processes) {
            if (!process->canSendMessage())
                continue;
            process->send(Messages::WebProcess::PostNotification(message), 0);
        }
    });
});
```

brentfulgham replied:
This should be efficient (and is based on all the work Chris Dumez did on removing `unsafeAppend`.

brentfulgham reviewed (approved):
Looks good. I had a suggestion for a cleaner/more efficient way to populate the notifications.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -784,6 +820,10 @@ static void logProcessPoolState(const WebProcessPool& pool)
 void WebProcessPool::unregisterNotificationObservers()
 {
+#if ENABLE(NOTIFYD_BLOCKING_IN_WEBCONTENT)
+    for (auto token : m_notifyTokens)
```
`int` is shorter than `auto` :)

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -645,6 +645,42 @@ static void logProcessPoolState(const WebProcessPool& pool)
+            for (auto& process : m_processes) {
+                if (!process->canSendMessage())
+                    continue;
+                process->send(Messages::WebProcess::PostNotification(message), 0);
```
This repeatedly constructs a WTF::String from a `const char*` for every process.

cdumez replied:
The following would be more efficient:
```
if (!m_processes.isEmpty()) {
    String messageString(message);
    for (auto& process : m_processes)
        process->send(Messages::WebProcess::PostNotification(messageString), 0);
}
```

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -645,6 +645,42 @@ static void logProcessPoolState(const WebProcessPool& pool)
+        int notifyToken;
+        notify_register_dispatch(message, &notifyToken, dispatch_get_main_queue(), ^(int token) {
+            for (auto& process : m_processes) {
+                if (!process->canSendMessage())
```
This check is not needed AFAIK

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1503,6 +1503,13 @@ static float currentBacklightLevel()
+#if ENABLE(NOTIFYD_BLOCKING_IN_WEBCONTENT)
+void WebProcess::postNotification(const String& message)
+{
+    notify_post(message.utf8().data());
```
Seems you could use `.ascii().data()`

pvollan replied:
Thanks for reviewing, all!

I will address these issues in the upcoming patch :)

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -645,6 +645,58 @@ static void logProcessPoolState(const WebProcessPool& pool)
+    };
+    m_notifyTokens = WTF::map(notificationMessages, [this](const ASCIILiteral& message) -> int {
+        int notifyToken;
+        notify_register_dispatch(message, &notifyToken, dispatch_get_main_queue(), ^(int token) {
```
`notify_register_dispatch` may fail. Maybe we should revise this code slightly to return a `std::optional<int>`, and return `nullopt` it the call fails.

Then we can switch from `WTF::map` to `WTF::compactMap`, which will return the `Vector` of all non-nullopt items.

(@cdumez Is my memory on how `compactMap` works right here?

cdumez replied:
Correct, if notify_register_dispatch() can fail, then we could switch to WTF::compactMap() and have the lambda return a `std::optional<int>`. Then you'd return std::nullopt in case of failure.

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -645,6 +645,58 @@ static void logProcessPoolState(const WebProcessPool& pool)
+    NSString* nsNotificationMessages[] = {
+        NSProcessInfoPowerStateDidChangeNotification
+    };
+    m_notificationObservers.reserveInitialCapacity(std::size(nsNotificationMessages));
```
Again, we should try to use the `WTF::compactMap` formulation here:

```
m_notificationObservers = WTF::compactMap(nsNotificationMessages, [this](NSString* message) -> RetainPtr<NSObject>  {
            RetainPtr<NSObject> observer = [[NSNotificationCenter defaultCenter] addObserverForName:message object:nil queue:[NSOperationQueue currentQueue] usingBlock:^(NSNotification *notification) {
            if (!m_processes.isEmpty()) {
                String messageString(message);
                for (auto& process : m_processes)
                    process->send(Messages::WebProcess::PostObserverNotification(message), 0);
            }
        }];
        return observer;
});
```

brentfulgham replied:
`compactMap` should return a Vector of all non-nil RetainPtrs.

pvollan replied:
Nice, thanks! I will address this in the upcoming patch :)

pvollan committed:
Send notifications from the UI process to the WebContent process
https://bugs.webkit.org/show_bug.cgi?id=263830
rdar://117629355

Reviewed by Brent Fulgham.

Send notifications from the UI process to the WebContent process. This is to prepare for blocking of notifyd in the WebContent process sandbox.

* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::postNotification):
(WebKit::WebProcess::postObserverNotification):

Canonical link: https://commits.webkit.org/270030@main
